### ISSUE TITLE
WalletTest failing on Github Actions ubuntu-latest

### ISSUE THREAD

**Issue by msgilligan** [bitcoinj/bitcoinj#2049]
See: [example](https://github.com/bitcoinj/bitcoinj/pull/2047/checks?check_run_id=1120297267)

This looks like it is related to `UnitTestParams.get()` returning a Genesis block with a `0` time via `genesisBlock.setTime(Utils.currentTimeSeconds())`

I'll look into this further and create a PR.

**Comment by schildbach** [comment]
I've not seen this for a while. Can this be closed?

**Comment by msgilligan** [comment]
Note that these failures have appeared again on **GitLab CI** using **Debian**. A PR is under development: PR #2147.

**Comment by schildbach** [comment]
The issue persists even with the `Block` refactorings of the last days.

**Comment by msgilligan** [comment]
> The issue persists even with the `Block` refactorings of the last days.

This doesn't surprise me because the refactoring has been fairly careful to not change the behavior. The two exceptions that I know about are:

1. The delayed construction of the genesis blocks (now that they're no longer created in the *NetParams constructors)
2. The slight increase in computing for Block.solve() now that the ~200 lowest bits in the uncompressed difficulty target are 0's where the were 1's.

I do think this refactoring makes the behavior of blocks much easier to reason about and will ultimately help us fix this issue or at least any related issues. Specifically, it will help us see exactly when (and for which type of blocks) "mock time" can override current time.

We do have a PR #2147 that fixes the current "occurrence" of this error, but I had the feeling that we were still playing "whack-a-mole" and that's why I started down the path of the `Block` refactoring. I think there will be one or two more commits on the refactoring and then it will be time to return to this issue and take a fresh look at PR #2147.

**Comment by msgilligan** [comment]
> 1. The delayed construction of the genesis blocks (now that they're no longer created in the *NetParams constructors)

This change seems fairly significant given that one factor in the behavior is that many tests contain the following line:

```
private static final NetworkParameters UNITTEST = UnitTestParams.get();
```

This means the creation time of the UNITTEST genesis block will be set by the first test that (likely indirectly) calls `.getGenesisBlock()` not by the first test class with the above line that is loaded.

(If we wanted to restore the old behavior, we could put a call to `.getGenesisBlock()` inside `UnitTestParams.get()`, but that shouldn't be necessary)

**Comment by msgilligan** [comment]
PR #2185 seems to fix the current manifestation of this issue. PR #2185 and PR #2186 combined should probably eliminated all occurrences of "block too far in the future" in our unit tests. They are both solid improvements and should be merged. They are sufficient to resolve this specific issue.

However, there may be other failures being caused by the MockClock. We can (and probably should) continue to fix them on a case-by-case basis, but we also need to think about medium-term and long-term solutions.

**Medium-term**:

We should either:

1. Reset the MockClock at the beginning of **every** unit test.
2. Reset the MockClock at the end of any unit test that activates the mock clock.

**Long-term**:

We should move away from production classes that use the `Utils` methods to get the current time and move towards mocking mechanisms that are only present in the test environment.

**Comment by schildbach** [comment]
I agree with your goals. Though I'd like to get rid of the mock clock sooner rather than later. Do you have a technical idea for your long-term goal?

**Comment by msgilligan** [comment]
> Do you have a technical idea for your long-term goal?

Do a major refactoring and rewrite of the majority of the unit tests (see Issue #1984) We would remove mutability from production classes and implement the tests via test mocking tools that generate/modify byte code. It would a be a major project, though it could be implemented over time.

We should also try (in an ongoing, case-by-case basis) to eliminate calls to the `Utils` time getter methods. For example, I think we can do that in the unit test genesis block (which means in all genesis blocks.)

I do recall finding some approaches with Google for mocking system time, so we could research that further/again.
